src/framework/matrix.hpp

```diff
@@ -533,6 +399,37 @@ template <class T> inline T matrix<T>::operator()(size_t i, size_t j) const {
   return mat_[j * rows_ + i];
 }
 
+// Addressing elements by row or column
+template <class T> inline std::vector<T> matrix<T>::row_index(size_t row) const {
+#ifdef DEBUG
+  if (row >= rows_) {
+    std::cerr << "Error: matrix class operator row_index out of bounds "
+              << row << " >= " << rows_ << std::endl;
+    exit(1);
+  }
+#endif
+  std::vector<T> ret;
+  ret.reserve(cols_);
+  for(size_t i = 0; i < cols_; i++)
+    ret.emplace_back(mat_[i * rows_ + row]);
+  return std::move(ret);
```

Comment:
I think we don't need to std::move here, the compiler will optimize away this return with a copy elision, which is even faster than move (as there's no move nor copy involved at all).

Reply:
I think you're right - I went a bit too far trying to fix the compiler warnings on the places where I'm returning py::dict/list but here I think the move is unnecessary because I'm sending a vector...?

Reply:
Yes, because the type we are returning is exactly the same as the function signature expects, there's copy elision. 
The other warnings you are referring to are correct, we shall m`std::move` there  because IIRC we were returning different (but compatible) types, so there's no copy elision, in this case a move will be better than copying (which is what happens by default).